services/npm/npm-last-update.service.js

```diff
@@ -0,0 +1,96 @@
+import Joi from 'joi'
+import dayjs from 'dayjs'
+import { optionalUrl } from '../validators.js'
+import {
+  BaseJsonService,
+  InvalidParameter,
+  pathParam,
+  queryParam,
+} from '../index.js'
+import { formatDate } from '../text-formatters.js'
+import { age as ageColor } from '../color-formatters.js'
+
+const queryParamSchema = Joi.object({
+  registry_uri: optionalUrl,
+  version: Joi.string(),
+}).required()
+
+const updateResponseSchema = Joi.object({
+  time: Joi.object({
+    created: Joi.string().required(),
+    modified: Joi.string().required(),
+  })
+    .pattern(Joi.string(), Joi.any())
+    .required(),
+}).required()
+
+const defaultUrl = 'https://registry.npmjs.org'
+
+export class NpmLastUpdate extends BaseJsonService {
```

Comment:
can we use `NpmBase` here? it has for example `fetchPackageData` that might be useful.

Reply:
- ` NpmBase`  appends to **/latest** to the package name. When `tag` variable is undefined.
- With `tag` not being undefined it points to correct url but not sure on the other effects.
- Please let me know if there is a better way, because of the above case I went with extending BaseJsonService


- also we’re using `version` query param which isn’t part of `NpmBase`

Reply:
Thank you for the clarification.

I see that `fetchPackageData` but we can still benefit from using `NpmBase`
Notice how we use `encodeScopedPackage` to take care of scoped packages. (which we might want here as well as a parameter for the badge)
Also using `_requestJson` from `NpmBase` could help us avoid https://github.com/npm/npmjs.org/issues/163

I think we will benefit from writing a fetch function for this use case that's separate from `fetchPackageData`, worst case it's clearer code, best case we will re-use it for other badges later.

Reply:
Got it Jonathan, let me give it a try again.

Reply:
Jonathan, extended `NpmBase` by adding a `fetch` function as suggested, please take a look and let me know if anyway we can make it better.